.taskcluster.yml

```diff
@@ -0,0 +1,71 @@
+version: 1
+reporting: checks-v1
+policy:
+  pullRequests: public
+tasks:
+  $let:
+    head_rev:
+      $if: tasks_for == "github-pull-request"
+      then: ${event.pull_request.head.sha}
+      else: ${event.after}
+    repository:
+      $if: tasks_for == "github-pull-request"
+      then: ${event.pull_request.head.repo.clone_url}
+      else: ${event.repository.url}
+  in:
+    $let:
+      tests:
+        - name: lint-ruby
+          image: ruby
+          command:
+            - /bin/bash
+            - '--login'
+            - '-c'
+            - >-
+              git clone ${repository} repo && cd repo &&
+              git config advice.detachedHead false &&
+              git checkout ${head_rev} &&
+              mv .gemrc ~/.gemrc &&
+              gem install rubocop &&
+              rubocop
+        - name: brew {audit, install}
+          image: ruby:2.6
+          command:
+            - /bin/bash
+            - '--login'
+            - '-c'
+            - >-
+              apt-get update && apt-get install -y build-essential procps file sudo &&
```

Comment:
Typically we build a docker image with the required toolchains, and then run the task in a container using that docker image, to reduce latency and CI overhead, rather than every task needing to perform the toolchain installations themselves.

See e.g. https://firefox-source-docs.mozilla.org/main/69.0/taskcluster/taskcluster/docker-images.html

I'm not sure if we have examples in GitHub of docker image builds based on this workflow, but I think it is a feature of taskgraph - @ahal might know?

In other words, if we use taskgraph, I think it should be able to take care of automatically building and indexing docker images used by CI tasks, so the repo just needs to contain the Dockerfile (e.g. which does all the apt-get installs) and when that is updated, taskgraph will take care of building the updated Docker images, updating the Index, and tasks that use it will automatically get the updated version. Then each task doesn't need to install toolchains etc, they are already part of the docker image.

Aside: we should probably introduce this workflow in the monorepo too. We don't install toolchains on each task run, which is good, but we don't have automated flows for rebuilding docker images when the Dockerfile changes, instead, we manually republish docker images to dockerhub. We would avoid this if we got taskgraph to do it for us, and the images would be task artifacts, rather than images on hub.docker.com.

Reply:
I've heard there are plans to get rid of docker worker.. but that will only happen when generic worker will support docker-worker-like payloads, so the images would still be used. 
Btw, in case of using generic-worker, you'd still need to bootstrap and install everything manually right? Or you would need to re-use some pre-built stuff from somewhere?